### DISCUSSION TITLE
QuarkusIntegrationTest : replacing 'Clock' bean with a fixed clock for tests

### DISCUSSION THREAD

**Discussion by kirankolli-addepar** [quarkusio/quarkus#39723] [Q&A]
This is a fairly common scenario - Applications use `clock` which is then referred to derive the current time. In case of `Integration Tests` we would like to replace this with a `Clock.fixed()` that points to a specific point in time instead of the regular clock that the application uses in dev or prod. How can we achieve this with Quarkus ? 

I have tried : 

Adding a bean with `IfBuildProfile` and `Priority(Integer.MAX_VALUE)` in the `QuarkusTestProfile` hoping that this bean would somehow replace the regular bean getting injected but that did not work.

Is the only way to override this is changing from `QuarkusIntegrationTest` to `QuarkusTest` ?

[a comment by an automated account is not shown]

**Comment by geoand** [comment]
`@QuarkusIntegrationTest` is meant to test the production application so it just runs the jar (or native binary or container), it cannot change the application in any way as `@QuarkusTest` does

**Reply by kirankolli-addepar** [reply]
ok, even in that case I tried putting this directly in the regular application config and with this I expect the fixed clock to be picked and yet it does not get picked : 

```
@ApplicationScoped
public class ClockProducer {

  @Produces
  @UnlessBuildProfile("test")
  public Clock clock() {
    return Clock.systemUTC();
  }

  @Produces
  @Priority(Integer.MAX_VALUE)
  @IfBuildProfile("test")
  public Clock testClock() {
    var testTime = LocalDateTime.of(2024, Month.FEBRUARY, 29, 10, 0).toInstant(ZoneOffset.UTC);
    return Clock.fixed(testTime, ZoneOffset.UTC);
  }
}
```

Shouldnt the test clock be picked while running integration tests ?

**Reply by geoand** [reply]
> Shouldnt the test clock be picked while running integration tests?

Nope, because integration-tests run with the prod profile

**Reply by kirankolli-addepar** [reply]
is there a way to add a `qualifier` to a bean in quarkus (similar to how its done in Spring) ?

**Reply by geoand** [reply]
https://quarkus.io/guides/cdi#you-talked-about-some-qualifiers

**Reply by kirankolli-addepar** [reply]
@geoand I just ran the integration test in debug mode and ran this during the test : 
`ConfigUtils.getProfiles()` and it returned the profile as 'test' . So Integration tests do not run in 'prod' profile.

**Reply by geoand** [reply]
I'd really like to see what you did, because that's not how it works

**Reply by kirankolli-addepar** [reply]
Answer to your question below. 

But in the mean time I have also tried changing the `QuarkusIntegrationTest` to a `QuarkusTest` so that I can inject a mock for the Clock.class. I did it like so 
```
@BeforeEach
  public void setupMocks() {
    ...
    var mockedClock = mock(Clock.class);
    when(mockedClock.instant()).thenReturn(testClock.instant());
    when(mockedClock.getZone()).thenReturn(testClock.getZone());
    QuarkusMock.installMockForType(mockedClock, Clock.class);
  }
```

And this keeps throwing this error below : 
```
java.time.Clock$SystemClock.arc$setMock(java.lang.Object)
java.lang.NoSuchMethodException: java.time.Clock$SystemClock.arc$setMock(java.lang.Object)
	at java.base/java.lang.Class.getDeclaredMethod(Class.java:2675)
	at io.quarkus.test.junit.MockSupport.installMock(MockSupport.java:53)
	at io.quarkus.test.junit.QuarkusMock.installMockForType(QuarkusMock.java:58)
```

Could this be because the original bean I am trying to replace is `Clock.systemUTC()` which is actually a static inner class inside java.time.Clock ?

**Reply by geoand** [reply]
> Could this be because the original bean I am trying to replace is Clock.systemUTC() which is actually a static inner class inside java.time.Clock ?

Yes, you won't be able to mock these
